api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+//
+// ExternalDNS describes a managed ExternalDNS controller instance for a cluster.
+// The controller is responsible for creating external DNS records in supported
+// DNS providers based off of instances of select Kubernetes resources.
+type ExternalDNS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-// ExternalDNSSpec defines the desired state of ExternalDNS
+	// spec is the specification of the desired behavior of the ExternalDNS.
+	Spec ExternalDNSSpec `json:"spec"`
+	// status is the most recently observed status of the ExternalDNS.
+	Status ExternalDNSStatus `json:"status"`
+}
+
+// ExternalDNSSpec defines the desired state of the ExternalDNS.
 type ExternalDNSSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Domains specifies which domains that ExternalDNS should
+	// create DNS records for. Multiple domain values
+	// can be specified such that subdomains of an included domain
+	// can effectively be ignored using the "Include" and "Exclude"
+	// domain filter options.
+	//
+	// An entry that excludes a domain that isn't a subdomain of some
+	// included domain will be ignored.
+	//
+	// An empty list of domains means ExternalDNS will create
+	// DNS records for any included source resource regardless
+	// of the resource's desired hostname.
+	//
+	// +optional
+	Domains []ExternalDNSDomain `json:"domains,omitempty"`
+
+	// Provider refers to the DNS provider that ExternalDNS
+	// should publish records to. Note that each ExternalDNS
+	// is tied to a single provider.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Provider ExternalDNSProvider `json:"provider"`
 
-	// Foo is an example field of ExternalDNS. Edit externaldns_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// Source describes which source resource
+	// ExternalDNS will be configured to create
+	// DNS records for.
+	//
+	// Multiple ExternalDNS CRs must be
+	// created if multiple ExternalDNS source resources
+	// are desired.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Source ExternalDNSSource `json:"source"`
+
+	// Zones describes which DNS Zone IDs
+	// ExternalDNS should publish records to.
+	//
+	// +optional
+	Zones []string `json:"zones,omitempty"`
```

Comment:
You have exceeded my DNS knowledge. If zones change, you probably have to react.  You need to describe the cleanup process.

Reply:
I've elaborated in the godocs. I think it makes sense for the operator to do nothing in this case.